test/python/result/test_mitigators.py

```diff
@@ -172,112 +56,175 @@ def compare_results(self, res1, res2):
+            np.array([[0.99023438, 0.02929688], [0.00976562, 0.97070312]]),
+            np.array([[0.984375, 0.02441406], [0.015625, 0.97558594]]),
+        ]
+        return assignment_matrices
+
+    @staticmethod
+    def counts_data(circuit, assignment_matrices, shots=1024):
+        """Generates count data for the noisy and noiseless versions of the circuit simulation"""
+        full_assignment_matrix = assignment_matrices[0]
+        for m in assignment_matrices[1:]:
+            full_assignment_matrix = np.kron(full_assignment_matrix, m)
+        num_qubits = len(assignment_matrices)
+        ideal_assignment_matrix = np.eye(2**num_qubits)
+        counts_ideal = TestReadoutMitigation.simulate_circuit(
+            circuit, ideal_assignment_matrix, num_qubits, shots
+        )
+        counts_noise = TestReadoutMitigation.simulate_circuit(
+            circuit, full_assignment_matrix, num_qubits, shots
+        )
+        probs_noise = {key: value / shots for key, value in counts_noise.items()}
+        return counts_ideal, counts_noise, probs_noise
+
+    def test_mitigation_improvement(self):
+        """Test whether readout mitigation led to more accurate results"""
+        shots = 1024
+        assignment_matrices = self.assignment_matrices()
+        num_qubits = len(assignment_matrices)
+        mitigators = self.mitigators(assignment_matrices)
+        circuit, circuit_name, num_qubits = self.ghz_3_circuit()
+        counts_ideal, counts_noise, probs_noise = self.counts_data(
+            circuit, assignment_matrices, shots
+        )
+        unmitigated_error = self.compare_results(counts_ideal, counts_noise)
+        unmitigated_stddev = stddev(probs_noise, shots)
+
+        for mitigator in mitigators:
+            mitigated_quasi_probs = mitigator.quasi_probabilities(counts_noise)
+            mitigated_probs = (
+                mitigated_quasi_probs.nearest_probability_distribution().binary_probabilities(
+                    num_bits=num_qubits
                 )
+            )
+            mitigated_error = self.compare_results(counts_ideal, mitigated_probs)
+            self.assertTrue(
+                mitigated_error < unmitigated_error * 0.8,
+                "Mitigator {} did not improve circuit {} measurements".format(
+                    mitigator, circuit_name
+                ),
+            )
+            mitigated_stddev_upper_bound = mitigated_quasi_probs._stddev_upper_bound
+            max_unmitigated_stddev = max(unmitigated_stddev.values())
+            self.assertTrue(
+                mitigated_stddev_upper_bound >= max_unmitigated_stddev,
+                "Mitigator {} on circuit {} gave stddev upper bound {} "
+                "while unmitigated stddev maximum is {}".format(
+                    mitigator,
+                    circuit_name,
+                    mitigated_stddev_upper_bound,
+                    max_unmitigated_stddev,
+                ),
+            )
```

Comment:
Generally (and elsewhere in the tests) it's better to use the more specific assertions like `self.assertGreaterEqual` because they'll automatically produce a good assertion message that includes the numbers.

The mitigators and circuit names would ideally be handled by `self.subTest`, but in practice we don't need to bother with that for this PR, because they don't work right with `testtools` unfortunately :(

Reply:
I agree, fixed in all occurances.